service/src/main/java/crawlercommons/urlfrontier/service/AbstractFrontierService.java

```diff
@@ -866,4 +867,78 @@ public void close() throws IOException {
     public abstract void getURLStatus(
             crawlercommons.urlfrontier.Urlfrontier.URLStatusRequest request,
             io.grpc.stub.StreamObserver<URLItem> responseObserver);
+
+    public void listURLs(
+            crawlercommons.urlfrontier.Urlfrontier.ListUrlParams request,
+            io.grpc.stub.StreamObserver<crawlercommons.urlfrontier.Urlfrontier.URLItem>
+                    responseObserver) {
+
+        long maxURLs = request.getSize();
+        long start = request.getStart();
+        String key = request.getKey();
+
+        final String normalisedCrawlID = CrawlID.normaliseCrawlID(request.getCrawlID());
+
+        // 100 by default
+        if (maxURLs == 0) {
+            maxURLs = 100;
+        }
+
+        LOG.info(
+                "Received request to list URLs [size {}; start {}; crawlId {}, key {}]",
+                maxURLs,
+                start,
+                normalisedCrawlID,
+                key);
+
+        synchronized (getQueues()) {
+            Iterator<Entry<QueueWithinCrawl, QueueInterface>> iterator =
+                    getQueues().entrySet().iterator();
+
+            while (iterator.hasNext()) {
+                Entry<QueueWithinCrawl, QueueInterface> e = iterator.next();
+
+                // check that it is within the right crawlID
+                if (!e.getKey().getCrawlid().equals(normalisedCrawlID)) {
+                    continue;
+                }
+
+                // check that it is within the right key/queue
+                if (key != null && !key.isEmpty() && !e.getKey().getQueue().equals(key)) {
+                    continue;
+                }
+                Iterator<URLItem> iter = urlIterator(e, start, maxURLs);
+                while (iter.hasNext()) {
+                    responseObserver.onNext(iter.next());
+                }
+            }
+        }
+
+        responseObserver.onCompleted();
+    }
+
+    protected abstract Iterator<URLItem> urlIterator(
+            Entry<QueueWithinCrawl, QueueInterface> qentry, long start, long max);
+
+    /**
+     * Builds an URLItem for listURLs (used by fetchURLItems, avoids builder instantiation for every
+     * URL)
+     *
+     * @param builder The URLItem builder
+     * @param kbuilder The KnownURLItem builder
+     * @param info URLInfo
+     * @param refetch refetch date from Epoch in seconds
+     * @return
+     */
+    public static URLItem buildURLItem(
+            URLItem.Builder builder, KnownURLItem.Builder kbuilder, URLInfo info, long refetch) {
```

Comment:
this is a nice way of simplifying the code in the sub classes + calling clear is also a safe way of making sure no data is carried through from a previous entry